internal/provider/waypoint/data_source_waypoint_application.go

```diff
@@ -0,0 +1,150 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package waypoint
+
+import (
+	"context"
+	"fmt"
+
+	sharedmodels "github.com/hashicorp/hcp-sdk-go/clients/cloud-shared/v1/models"
+	waypoint_models "github.com/hashicorp/hcp-sdk-go/clients/cloud-waypoint-service/preview/2023-08-18/models"
+	"github.com/hashicorp/terraform-plugin-framework-validators/datasourcevalidator"
+	"github.com/hashicorp/terraform-plugin-framework/datasource"
+	"github.com/hashicorp/terraform-plugin-framework/datasource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/path"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/hashicorp/terraform-provider-hcp/internal/clients"
+)
+
+var _ datasource.DataSource = &DataSourceApplication{}
+var _ datasource.DataSourceWithConfigValidators = &DataSourceApplication{}
+
+func (d DataSourceApplication) ConfigValidators(ctx context.Context) []datasource.ConfigValidator {
+	return []datasource.ConfigValidator{
+		datasourcevalidator.Conflicting(
+			path.MatchRoot("name"),
+			path.MatchRoot("id"),
+		),
+	}
+}
+
+type DataSourceApplication struct {
+	client *clients.Client
+}
+
+func NewApplicationDataSource() datasource.DataSource {
+	return &DataSourceApplication{}
+}
+
+func (d *DataSourceApplication) Metadata(ctx context.Context, req datasource.MetadataRequest, resp *datasource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_waypoint_application"
+}
+
+func (d *DataSourceApplication) Schema(ctx context.Context, req datasource.SchemaRequest, resp *datasource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		MarkdownDescription: "The Waypoint Application data source retrieves information on a given Application.",
+		Attributes: map[string]schema.Attribute{
+			"id": schema.StringAttribute{
+				Computed:    true,
+				Optional:    true,
+				Description: "The ID of the Application.",
+			},
+			"name": schema.StringAttribute{
+				Description: "The name of the Application.",
+				Computed:    true,
+				Optional:    true,
+			},
+			"organization_id": schema.StringAttribute{
+				Description: "The ID of the HCP organization where the Waypoint Application is located.",
+				Computed:    true,
+			},
+			"project_id": schema.StringAttribute{
+				Description: "The ID of the HCP project where the Waypoint Application is located.",
+				Optional:    true,
+				Computed:    true,
+			},
+			"readme_markdown": schema.StringAttribute{
+				Computed:    true,
+				Description: "Instructions for using the Application (markdown format supported).",
+			},
+			"application_template_id": schema.StringAttribute{
+				Computed:    true,
+				Description: "ID of the Application Template this Application is based on.",
+			},
+			"application_template_name": schema.StringAttribute{
+				Computed:    true,
+				Description: "Name of the Application Template this Application is based on.",
+			},
```

Comment:
I think @paladin-devops and I had this conversation over on the HCP CLI, but it sounds like we're just calling this `template` instead of `application template`?

Reply:
Yeah, I chatted with @anubhavmishra recently who told me that with customers, he refers to these just as "templates" (after I retold the story of how we used to have project templates vs. application templates). The CLI command for this resource is thusly just "templates" instead of "applications templates". The UI refers to them as "templates" too.

Reply:
Well that's a bit of a change considering we already have a resource called `application_template`. I'm not strongly opposed to renaming, but I'd rather do that in a separate PR that does it everywhere